src/Interpreters/Aggregator.cpp

```diff
@@ -175,6 +194,77 @@ Block Aggregator::getHeader(bool final) const
     return params.getHeader(header, final);
 }
 
+Aggregator::Params::Params(
+    const Settings & settings,
+    const Names & keys_,
+    const AggregateDescriptions & aggregates_,
+    bool overflow_row_,
+    size_t group_by_two_level_threshold_,
+    size_t group_by_two_level_threshold_bytes_,
+    bool empty_result_for_aggregation_by_empty_set_,
+    TemporaryDataOnDiskScopePtr tmp_data_scope_,
+    bool only_merge_, // true for projections
+    const StatsCollectingParams & stats_collecting_params_)
+    : keys(keys_)
+    , keys_size(keys.size())
+    , aggregates(aggregates_)
+    , aggregates_size(aggregates.size())
+    , overflow_row(overflow_row_)
+    , max_rows_to_group_by(settings[Setting::max_rows_to_group_by])
+    , group_by_overflow_mode(settings[Setting::group_by_overflow_mode])
+    , group_by_two_level_threshold(group_by_two_level_threshold_)
+    , group_by_two_level_threshold_bytes(group_by_two_level_threshold_bytes_)
+    , max_bytes_before_external_group_by(settings[Setting::max_bytes_before_external_group_by])
+    , empty_result_for_aggregation_by_empty_set(empty_result_for_aggregation_by_empty_set_)
+    , tmp_data_scope(std::move(tmp_data_scope_))
+    , max_threads(settings[Setting::max_threads])
+    , min_free_disk_space(settings[Setting::min_free_disk_space_for_temporary_data])
+    , compile_aggregate_expressions(settings[Setting::compile_aggregate_expressions])
+    , min_count_to_compile_aggregate_expression(settings[Setting::min_count_to_compile_aggregate_expression])
+    , max_block_size(settings[Setting::max_block_size])
+    , only_merge(only_merge_)
+    , enable_prefetch(settings[Setting::enable_software_prefetch_in_aggregation])
+    , optimize_group_by_constant_keys(settings[Setting::optimize_group_by_constant_keys])
+    , min_hit_rate_to_use_consecutive_keys_optimization(settings[Setting::min_hit_rate_to_use_consecutive_keys_optimization])
+    , stats_collecting_params(stats_collecting_params_)
+{
+    if (settings[Setting::max_bytes_ratio_before_external_group_by] != 0.)
+    {
+        if (settings[Setting::max_bytes_before_external_group_by] > 0)
+            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Settings max_bytes_ratio_before_external_group_by and max_bytes_before_external_group_by cannot be set simultaneously");
+
+        double ratio = settings[Setting::max_bytes_ratio_before_external_group_by];
+        if (ratio < 0 || ratio >= 1.)
+            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Setting max_bytes_ratio_before_external_group_by should be >= 0 and < 1 ({})", ratio);
+
+        UInt64 available_system_memory = getMostStrictAvailableSystemMemory();
+        max_bytes_before_external_group_by = static_cast<size_t>(available_system_memory * ratio);
```

Comment:
maybe it is better to not rely on uint64_t max here and if there is no limit on memory just log something and ignore this setting?

Reply:
Make sense, thanks, fixed.